Makefile

```
@@ -1,50 +1,87 @@
+# Convenience data so we can run the following and include
+# sources from three directories deep.
+#
+# Example:
+# $(subst dir, emily/cdk/lib, $(THREE_DIRS_DEEP))
+# becomes
+# emily/cdk/lib/*  emily/cdk/lib/*/*  emily/cdk/lib/*/*/*
+THREE_DIRS_DEEP := dir/* dir/*/* dir/*/*/*
+TWO_DIRS_DEEP := dir/* dir/*/*
+
+AUTOGENERATED_EMILY_CLIENT := .generated-sources/emily/src/lib.rs
+EMILY_LAMBDA_BINARY := target/lambda/emily-lambda/bootstrap.zip
+EMILY_CDK_TEMPLATE := emily/cdk/cdk.out/EmilyStack.template.json
+EMILY_DOCKER_COMPOSE := docker-compose.emily.yml
+INSTALL_TARGET := pnpm-lock.yaml
+
+# Don't use the install target here so you can rerun install without
+# Makefile complaints.
 install:
 	pnpm install
+	touch pnpm-lock.yaml
 
-build: install | emily-client-source
+build: $(INSTALL_TARGET) $(AUTOGENERATED_EMILY_CLIENT)
 	cargo build
 	pnpm --recursive build
 
-test: install | emily-client-source
+test: $(INSTALL_TARGET) $(AUTOGENERATED_EMILY_CLIENT)
 	cargo test
 	pnpm --recursive test
 
-integration-test: install | emily-client-source
+integration-test: $(INSTALL_TARGET) $(AUTOGENERATED_EMILY_CLIENT)
 	docker compose --file docker-compose.test.yml up --detach
 	cargo test --all-features
 	pnpm --recursive test
 	docker compose --file docker-compose.test.yml down
 
-lint: install | emily-client-source
+lint: $(INSTALL_TARGET) $(AUTOGENERATED_EMILY_CLIENT)
 	cargo clippy -- -D warnings
 	pnpm --recursive run lint
 
 clean:
 	cargo clean
 	pnpm --recursive clean
+	rm -rf devenv/dynamodb/data/*
+	@touch package.json
+
+.PHONY: install build test integration-test lint clean
+
+$(INSTALL_TARGET): $(wildcard package* */package* */*/package*)
+	pnpm install
+	touch pnpm-lock.yaml
 
 # Emily API
 # ----------------------------------------------------
 
-EMILY_PATH=emily
+EMILY_DEVENV_PATH=emily/devenv
+EMILY_LAMBDA_PATH=emily/lambda
 EMILY_API_PROJECT_NAME=emily-api
 EMILY_CDK_PROJECT_NAME=emily-cdk
 
-emily-cdk: emily-client | emily-lambda
-	pnpm --filter $(EMILY_CDK_PROJECT_NAME) run build
+# Launches Emily dev environment.
+emily-integration-test: devenv $(EMILY_LAMBDA_BINARY) $(EMILY_CDK_TEMPLATE) $(EMILY_DOCKER_COMPOSE)
+	docker compose --file docker-compose.emily.yml up --remove-orphans
+.PHONY: emily-integration-test
 
-# Overwrite the default architecture with --arm64 configuration
-# on arm machines.
-ifneq (filter arm64 aarch64, $(shell uname -m),)
+# Builds all dockerfiles that need to be built for the dev environment.
+devenv: $(wildcard $(subst dir, devenv, $(TWO_DIRS_DEEP)))
+	sudo docker compose -f docker-compose.emily.yml build
```

I would prefer not having our build scripts require superuser privileges.

```suggestion
	docker compose -f docker-compose.emily.yml build
```

---

SGTM

---

SGTM?

---

sounds good to me

---

Aaah, obviously :sweat_smile: